server/lib/src/idm/accountpolicy.rs

```diff
@@ -0,0 +1,153 @@
+use crate::prelude::*;
+// use crate::idm::server::IdmServerProxyWriteTransaction;
+
+#[derive(Copy, Clone, Debug, Ord, PartialOrd, Eq, PartialEq, Default)]
+#[repr(u32)]
+pub(crate) enum CredentialPolicy {
+    #[default]
+    NoPolicy = 0,
+    MfaRequired = 10,
+    PasskeyRequired = 20,
+    AttestedPasskeyRequired = 30,
+    AttestedResidentKeyRequired = 40,
+}
+
+impl From<u32> for CredentialPolicy {
+    fn from(value: u32) -> Self {
+        if value >= CredentialPolicy::AttestedResidentKeyRequired as u32 {
+            CredentialPolicy::AttestedResidentKeyRequired
+        } else if value >= CredentialPolicy::AttestedPasskeyRequired as u32 {
+            CredentialPolicy::AttestedPasskeyRequired
+        } else if value >= CredentialPolicy::PasskeyRequired as u32 {
+            CredentialPolicy::PasskeyRequired
+        } else if value >= CredentialPolicy::MfaRequired as u32 {
+            CredentialPolicy::MfaRequired
+        } else {
+            CredentialPolicy::NoPolicy
+        }
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct AccountPolicy {
+    privilege_expiry: u32,
+    authsession_expiry: u32,
+    credential_policy: CredentialPolicy,
+}
+
+impl Into<Option<AccountPolicy>> for &EntrySealedCommitted {
+    fn into(self) -> Option<AccountPolicy> {
+        if !self.attribute_equality(
+            Attribute::Class,
+            &EntryClass::AccountPolicy.to_partialvalue(),
+        ) {
+            return None;
+        }
+
+        let authsession_expiry = self
+            .get_ava_single_uint32(Attribute::AuthSessionExpiry)
+            .unwrap_or(MAXIMUM_AUTH_SESSION_EXPIRY);
+        let privilege_expiry = self
+            .get_ava_single_uint32(Attribute::PrivilegeExpiry)
+            .unwrap_or(MAXIMUM_AUTH_PRIVILEGE_EXPIRY);
+        let credential_policy = CredentialPolicy::default();
+
+        Some(AccountPolicy {
+            privilege_expiry,
+            authsession_expiry,
+            credential_policy,
+        })
+    }
+}
+
+#[derive(Clone)]
+#[cfg_attr(test, derive(Default))]
+pub(crate) struct ResolvedAccountPolicy {
+    privilege_expiry: u32,
+    authsession_expiry: u32,
+    credential_policy: CredentialPolicy,
+}
+
+impl ResolvedAccountPolicy {
+    pub(crate) fn fold_from<I>(iter: I) -> Self
+    where
+        I: Iterator<Item = AccountPolicy>,
+    {
+        // Start with our maximums
+        let mut accumulate = ResolvedAccountPolicy {
+            privilege_expiry: MAXIMUM_AUTH_PRIVILEGE_EXPIRY,
+            authsession_expiry: MAXIMUM_AUTH_SESSION_EXPIRY,
+            credential_policy: CredentialPolicy::default(),
+        };
+
+        iter.for_each(|acc_pol| {
+            // Take the smaller expiry
+            if acc_pol.privilege_expiry < accumulate.privilege_expiry {
+                accumulate.privilege_expiry = acc_pol.privilege_expiry
+            }
```

Comment:
`itertools::min/max` is pretty rad for the iterator, would allow a big collapse of the lewp:
```
        let min_priv_expiry = itertools::min(acc_pol.iter().map(|a| a. privilege_expiry));

        // define the policy
        let mut accumulate = ResolvedAccountPolicy {
            privilege_expiry: std::cmp::min(MAXIMUM_AUTH_PRIVILEGE_EXPIRY, min_priv_expiry),
            authsession_expiry: MAXIMUM_AUTH_SESSION_EXPIRY,
            credential_policy: CredentialPolicy::default(),
        };

      
```

Reply:
Yeah, but we can only go over the iterator *once*. We can't do this because it requires us to iterate over the iter multiple times.

Reply:
Ah, because we're doing a from on the parent iter I guess. Boooooo.